### PR TITLE
Fix duplicated UUID migration issue

### LINKED ISSUE #30936: [Bug]: [user_ldap] Migration from 1.13.0 to 1.13.1 broken
### ⚠️ This issue respects the following points: ⚠️

- [X] This is a **bug**, not a question or a configuration/webserver/proxy issue.
- [X] This issue is **not** already reported on Github _(I've searched it)_.
- [X] Nextcloud Server **is** up to date. See [Maintenance and Release Schedule](https://github.com/nextcloud/server/wiki/Maintenance-and-Release-Schedule) for supported versions.
- [X] I agree to follow Nextcloud's [Code of Conduct](https://nextcloud.com/contribute/code-of-conduct/).

### Bug description

Migration of user_ldap app fails:
`Database error when running migration latest for app user_ldap`

### Steps to reproduce

1. Upgrade NC 23.0.0.10 to 23.0.1.0
2. Upgrade LDAP App from 1.13.0 to 1.13.1

### Expected behavior

No issues while upgrading the app.

### Installation method

Other

### Operating system

Other

### PHP engine version

PHP 7.4

### Web server

Nginx

### Database engine version

MariaDB

### Is this bug present after an update or on a fresh install?

Updated from a minor version (ex. 22.2.3 to 22.2.4)

### Are you using the Nextcloud Server Encryption module?

Encryption is Disabled

### What user-backends are you using?

- [X] Default user-backend _(database)_
- [X] LDAP/ Active Directory
- [ ] SSO - SAML
- [ ] Other

### Configuration report

```shell
Nextcloud or one of the apps require upgrade - only a limited number of commands are available
You may use your browser or the occ upgrade command to do the upgrade
{
    "system": {
        "instanceid": "***REMOVED SENSITIVE VALUE***",
        "passwordsalt": "***REMOVED SENSITIVE VALUE***",
        "secret": "***REMOVED SENSITIVE VALUE***",
        "trusted_proxies": "***REMOVED SENSITIVE VALUE***",
        "overwritehost": "nextcloud.domain.net",
        "overwriteprotocol": "https",
        "trusted_domains": [
            "nextcloud.domain.net"
        ],
        "datadirectory": "***REMOVED SENSITIVE VALUE***",
        "overwrite.cli.url": "https:\/\/nextcloud.domain.net",
        "default_language": "de",
        "dbtype": "mysql",
        "version": "23.0.0.10",
        "dbname": "***REMOVED SENSITIVE VALUE***",
        "dbhost": "***REMOVED SENSITIVE VALUE***",
        "dbport": "",
        "dbtableprefix": "oc_",
        "dbuser": "***REMOVED SENSITIVE VALUE***",
        "dbpassword": "***REMOVED SENSITIVE VALUE***",
        "logtimezone": "UTC",
        "loglevel": "2",
        "installed": true,
        "memcache.local": "\\OC\\Memcache\\APCu",
        "memcache.locking": "\\OC\\Memcache\\Redis",
        "redis": {
            "host": "***REMOVED SENSITIVE VALUE***",
            "port": 6379
        },
        "ldapIgnoreNamingRules": false,
        "ldapProviderFactory": "\\OCA\\User_LDAP\\LDAPProviderFactory",
        "maintenance": false,
        "theme": "",
        "mail_from_address": "***REMOVED SENSITIVE VALUE***",
        "mail_smtpmode": "smtp",
        "mail_domain": "***REMOVED SENSITIVE VALUE***",
        "updater.release.channel": "stable",
        "mail_smtphost": "***REMOVED SENSITIVE VALUE***",
        "default_phone_region": "DE",
        "app_install_overwrite": [
            "quota_warning"
        ],
        "mysql.utf8mb4": true
    }
}
```

### List of activated Apps

```shell
Nextcloud or one of the apps require upgrade - only a limited number of commands are available
You may use your browser or the occ upgrade command to do the upgrade
Enabled:
  - accessibility: 1.9.0
  - activity: 2.15.0
  - cloud_federation_api: 1.6.0
  - comments: 1.13.0
  - contactsinteraction: 1.4.0
  - dashboard: 7.3.0
  - dav: 1.21.0
  - federatedfilesharing: 1.13.0
  - federation: 1.13.0
  - files: 1.18.0
  - files_pdfviewer: 2.4.0
  - files_rightclick: 1.2.0
  - files_sharing: 1.15.0
  - files_trashbin: 1.13.0
  - files_versions: 1.16.0
  - files_videoplayer: 1.12.0
  - firstrunwizard: 2.12.0
  - logreader: 2.8.0
  - lookup_server_connector: 1.11.0
  - nextcloud_announcements: 1.12.0
  - notifications: 2.11.1
  - oauth2: 1.11.0
  - password_policy: 1.13.0
  - photos: 1.5.0
  - privacy: 1.7.0
  - provisioning_api: 1.13.0
  - recommendations: 1.2.0
  - serverinfo: 1.13.0
  - settings: 1.5.0
  - sharebymail: 1.13.0
  - support: 1.6.0
  - survey_client: 1.11.0
  - systemtags: 1.13.0
  - text: 3.4.0
  - theming: 1.14.0
  - twofactor_backupcodes: 1.12.0
  - updatenotification: 1.13.0
  - user_ldap: 1.13.0
  - user_status: 1.3.1
  - viewer: 1.7.0
  - weather_status: 1.3.0
  - workflowengine: 2.5.0
Disabled:
  - admin_audit
  - circles: 0.19.5
  - encryption
  - files_external
```

### Nextcloud Signing status

```shell
No errors have been found.
```

### Nextcloud Logs

```shell
{"reqId":"b2XaK6tsG7sBep1nKa2E","level":3,"time":"2022-01-31T09:47:54+00:00","remoteAddr":"","user":"--","app":"no app in context","method":"","url":"--","message":"Database error when running migration latest for app user_ldap","userAgent":"--","version":"23.0.0.10","exception":{"Exception":"Exception","Message":"Database error when running migration latest for app user_ldap","Code":0,"Trace":[{"file":"/var/www/html/lib/private/legacy/OC_App.php","line":1011,"function":"migrate","class":"OC\\DB\\MigrationService","type":"->","args":[]},{"file":"/var/www/html/lib/private/Updater.php","line":353,"function":"updateApp","class":"OC_App","type":"::","args":["*** sensitive parameters replaced ***"]},{"file":"/var/www/html/lib/private/Updater.php","line":265,"function":"doAppUpgrade","class":"OC\\Updater","type":"->","args":[]},{"file":"/var/www/html/lib/private/Updater.php","line":133,"function":"doUpgrade","class":"OC\\Updater","type":"->","args":["23.0.1.2","23.0.0.10"]},{"file":"/var/www/html/core/Command/Upgrade.php","line":235,"function":"upgrade","class":"OC\\Updater","type":"->","args":[]},{"file":"/var/www/html/3rdparty/symfony/console/Command/Command.php","line":255,"function":"execute","class":"OC\\Core\\Command\\Upgrade","type":"->","args":[{"__class__":"Symfony\\Component\\Console\\Input\\ArgvInput"},{"__class__":"Symfony\\Component\\Console\\Output\\ConsoleOutput"}]},{"file":"/var/www/html/3rdparty/symfony/console/Application.php","line":1009,"function":"run","class":"Symfony\\Component\\Console\\Command\\Command","type":"->","args":[{"__class__":"Symfony\\Component\\Console\\Input\\ArgvInput"},{"__class__":"Symfony\\Component\\Console\\Output\\ConsoleOutput"}]},{"file":"/var/www/html/3rdparty/symfony/console/Application.php","line":273,"function":"doRunCommand","class":"Symfony\\Component\\Console\\Application","type":"->","args":[{"__class__":"OC\\Core\\Command\\Upgrade"},{"__class__":"Symfony\\Component\\Console\\Input\\ArgvInput"},{"__class__":"Symfony\\Component\\Console\\Output\\ConsoleOutput"}]},{"file":"/var/www/html/3rdparty/symfony/console/Application.php","line":149,"function":"doRun","class":"Symfony\\Component\\Console\\Application","type":"->","args":[{"__class__":"Symfony\\Component\\Console\\Input\\ArgvInput"},{"__class__":"Symfony\\Component\\Console\\Output\\ConsoleOutput"}]},{"file":"/var/www/html/lib/private/Console/Application.php","line":211,"function":"run","class":"Symfony\\Component\\Console\\Application","type":"->","args":[{"__class__":"Symfony\\Component\\Console\\Input\\ArgvInput"},{"__class__":"Symfony\\Component\\Console\\Output\\ConsoleOutput"}]},{"file":"/var/www/html/console.php","line":99,"function":"run","class":"OC\\Console\\Application","type":"->","args":[]},{"file":"/var/www/html/occ","line":11,"args":["/var/www/html/console.php"],"function":"require_once"}],"File":"/var/www/html/lib/private/DB/MigrationService.php","Line":429,"Previous":{"Exception":"Doctrine\\DBAL\\Exception\\UniqueConstraintViolationException","Message":"An exception occurred while executing a query: SQLSTATE[23000]: Integrity constraint violation: 1062 Duplicate entry '300' for key 'ldap_user_directory_uuid'","Code":1062,"Trace":[{"file":"/var/www/html/3rdparty/doctrine/dbal/src/Connection.php","line":1780,"function":"convert","class":"Doctrine\\DBAL\\Driver\\API\\MySQL\\ExceptionConverter","type":"->","args":[{"__class__":"Doctrine\\DBAL\\Driver\\PDO\\Exception"},{"__class__":"Doctrine\\DBAL\\Query"}]},{"file":"/var/www/html/3rdparty/doctrine/dbal/src/Connection.php","line":1719,"function":"handleDriverException","class":"Doctrine\\DBAL\\Connection","type":"->","args":[{"__class__":"Doctrine\\DBAL\\Driver\\PDO\\Exception"},{"__class__":"Doctrine\\DBAL\\Query"}]},{"file":"/var/www/html/3rdparty/doctrine/dbal/src/Connection.php","line":1067,"function":"convertExceptionDuringQuery","class":"Doctrine\\DBAL\\Connection","type":"->","args":[{"__class__":"Doctrine\\DBAL\\Driver\\PDO\\Exception"},"CREATE UNIQUE INDEX ldap_user_directory_uuid ON oc_ldap_user_mapping (directory_uuid)",[],[]]},{"file":"/var/www/html/lib/private/DB/Connection.php","line":237,"function":"executeQuery","class":"Doctrine\\DBAL\\Connection","type":"->","args":["CREATE UNIQUE INDEX ldap_user_directory_uuid ON oc_ldap_user_mapping (directory_uuid)",[],[],null]},{"file":"/var/www/html/3rdparty/doctrine/dbal/src/Connection.php","line":1809,"function":"executeQuery","class":"OC\\DB\\Connection","type":"->","args":["CREATE UNIQUE INDEX ldap_user_directory_uuid ON oc_ldap_user_mapping (directory_uuid)"]},{"file":"/var/www/html/lib/private/DB/Migrator.php","line":175,"function":"query","class":"Doctrine\\DBAL\\Connection","type":"->","args":["CREATE UNIQUE INDEX ldap_user_directory_uuid ON oc_ldap_user_mapping (directory_uuid)"]},{"file":"/var/www/html/lib/private/DB/Migrator.php","line":76,"function":"applySchema","class":"OC\\DB\\Migrator","type":"->","args":[{"__class__":"Doctrine\\DBAL\\Schema\\Schema"}]},{"file":"/var/www/html/lib/private/DB/Connection.php","line":546,"function":"migrate","class":"OC\\DB\\Migrator","type":"->","args":[{"__class__":"Doctrine\\DBAL\\Schema\\Schema"}]},{"file":"/var/www/html/lib/private/DB/MigrationService.php","line":540,"function":"migrateToSchema","class":"OC\\DB\\Connection","type":"->","args":[{"__class__":"Doctrine\\DBAL\\Schema\\Schema"}]},{"file":"/var/www/html/lib/private/DB/MigrationService.php","line":425,"function":"executeStep","class":"OC\\DB\\MigrationService","type":"->","args":["1130Date20211102154716",false]},{"file":"/var/www/html/lib/private/legacy/OC_App.php","line":1011,"function":"migrate","class":"OC\\DB\\MigrationService","type":"->","args":[]},{"file":"/var/www/html/lib/private/Updater.php","line":353,"function":"updateApp","class":"OC_App","type":"::","args":["*** sensitive parameters replaced ***"]},{"file":"/var/www/html/lib/private/Updater.php","line":265,"function":"doAppUpgrade","class":"OC\\Updater","type":"->","args":[]},{"file":"/var/www/html/lib/private/Updater.php","line":133,"function":"doUpgrade","class":"OC\\Updater","type":"->","args":["23.0.1.2","23.0.0.10"]},{"file":"/var/www/html/core/Command/Upgrade.php","line":235,"function":"upgrade","class":"OC\\Updater","type":"->","args":[]},{"file":"/var/www/html/3rdparty/symfony/console/Command/Command.php","line":255,"function":"execute","class":"OC\\Core\\Command\\Upgrade","type":"->","args":[{"__class__":"Symfony\\Component\\Console\\Input\\ArgvInput"},{"__class__":"Symfony\\Component\\Console\\Output\\ConsoleOutput"}]},{"file":"/var/www/html/3rdparty/symfony/console/Application.php","line":1009,"function":"run","class":"Symfony\\Component\\Console\\Command\\Command","type":"->","args":[{"__class__":"Symfony\\Component\\Console\\Input\\ArgvInput"},{"__class__":"Symfony\\Component\\Console\\Output\\ConsoleOutput"}]},{"file":"/var/www/html/3rdparty/symfony/console/Application.php","line":273,"function":"doRunCommand","class":"Symfony\\Component\\Console\\Application","type":"->","args":[{"__class__":"OC\\Core\\Command\\Upgrade"},{"__class__":"Symfony\\Component\\Console\\Input\\ArgvInput"},{"__class__":"Symfony\\Component\\Console\\Output\\ConsoleOutput"}]},{"file":"/var/www/html/3rdparty/symfony/console/Application.php","line":149,"function":"doRun","class":"Symfony\\Component\\Console\\Application","type":"->","args":[{"__class__":"Symfony\\Component\\Console\\Input\\ArgvInput"},{"__class__":"Symfony\\Component\\Console\\Output\\ConsoleOutput"}]},{"file":"/var/www/html/lib/private/Console/Application.php","line":211,"function":"run","class":"Symfony\\Component\\Console\\Application","type":"->","args":[{"__class__":"Symfony\\Component\\Console\\Input\\ArgvInput"},{"__class__":"Symfony\\Component\\Console\\Output\\ConsoleOutput"}]},{"file":"/var/www/html/console.php","line":99,"function":"run","class":"OC\\Console\\Application","type":"->","args":[]},{"file":"/var/www/html/occ","line":11,"args":["/var/www/html/console.php"],"function":"require_once"}],"File":"/var/www/html/3rdparty/doctrine/dbal/src/Driver/API/MySQL/ExceptionConverter.php","Line":60,"Previous":{"Exception":"Doctrine\\DBAL\\Driver\\PDO\\Exception","Message":"SQLSTATE[23000]: Integrity constraint violation: 1062 Duplicate entry '300' for key 'ldap_user_directory_uuid'","Code":1062,"Trace":[{"file":"/var/www/html/3rdparty/doctrine/dbal/src/Driver/PDO/Connection.php","line":87,"function":"new","class":"Doctrine\\DBAL\\Driver\\PDO\\Exception","type":"::","args":[{"__class__":"PDOException","errorInfo":["23000",1062,"Duplicate entry '300' for key 'ldap_user_directory_uuid'"]}]},{"file":"/var/www/html/3rdparty/doctrine/dbal/src/Connection.php","line":1062,"function":"query","class":"Doctrine\\DBAL\\Driver\\PDO\\Connection","type":"->","args":["CREATE UNIQUE INDEX ldap_user_directory_uuid ON oc_ldap_user_mapping (directory_uuid)"]},{"file":"/var/www/html/lib/private/DB/Connection.php","line":237,"function":"executeQuery","class":"Doctrine\\DBAL\\Connection","type":"->","args":["CREATE UNIQUE INDEX ldap_user_directory_uuid ON oc_ldap_user_mapping (directory_uuid)",[],[],null]},{"file":"/var/www/html/3rdparty/doctrine/dbal/src/Connection.php","line":1809,"function":"executeQuery","class":"OC\\DB\\Connection","type":"->","args":["CREATE UNIQUE INDEX ldap_user_directory_uuid ON oc_ldap_user_mapping (directory_uuid)"]},{"file":"/var/www/html/lib/private/DB/Migrator.php","line":175,"function":"query","class":"Doctrine\\DBAL\\Connection","type":"->","args":["CREATE UNIQUE INDEX ldap_user_directory_uuid ON oc_ldap_user_mapping (directory_uuid)"]},{"file":"/var/www/html/lib/private/DB/Migrator.php","line":76,"function":"applySchema","class":"OC\\DB\\Migrator","type":"->","args":[{"__class__":"Doctrine\\DBAL\\Schema\\Schema"}]},{"file":"/var/www/html/lib/private/DB/Connection.php","line":546,"function":"migrate","class":"OC\\DB\\Migrator","type":"->","args":[{"__class__":"Doctrine\\DBAL\\Schema\\Schema"}]},{"file":"/var/www/html/lib/private/DB/MigrationService.php","line":540,"function":"migrateToSchema","class":"OC\\DB\\Connection","type":"->","args":[{"__class__":"Doctrine\\DBAL\\Schema\\Schema"}]},{"file":"/var/www/html/lib/private/DB/MigrationService.php","line":425,"function":"executeStep","class":"OC\\DB\\MigrationService","type":"->","args":["1130Date20211102154716",false]},{"file":"/var/www/html/lib/private/legacy/OC_App.php","line":1011,"function":"migrate","class":"OC\\DB\\MigrationService","type":"->","args":[]},{"file":"/var/www/html/lib/private/Updater.php","line":353,"function":"updateApp","class":"OC_App","type":"::","args":["*** sensitive parameters replaced ***"]},{"file":"/var/www/html/lib/private/Updater.php","line":265,"function":"doAppUpgrade","class":"OC\\Updater","type":"->","args":[]},{"file":"/var/www/html/lib/private/Updater.php","line":133,"function":"doUpgrade","class":"OC\\Updater","type":"->","args":["23.0.1.2","23.0.0.10"]},{"file":"/var/www/html/core/Command/Upgrade.php","line":235,"function":"upgrade","class":"OC\\Updater","type":"->","args":[]},{"file":"/var/www/html/3rdparty/symfony/console/Command/Command.php","line":255,"function":"execute","class":"OC\\Core\\Command\\Upgrade","type":"->","args":[{"__class__":"Symfony\\Component\\Console\\Input\\ArgvInput"},{"__class__":"Symfony\\Component\\Console\\Output\\ConsoleOutput"}]},{"file":"/var/www/html/3rdparty/symfony/console/Application.php","line":1009,"function":"run","class":"Symfony\\Component\\Console\\Command\\Command","type":"->","args":[{"__class__":"Symfony\\Component\\Console\\Input\\ArgvInput"},{"__class__":"Symfony\\Component\\Console\\Output\\ConsoleOutput"}]},{"file":"/var/www/html/3rdparty/symfony/console/Application.php","line":273,"function":"doRunCommand","class":"Symfony\\Component\\Console\\Application","type":"->","args":[{"__class__":"OC\\Core\\Command\\Upgrade"},{"__class__":"Symfony\\Component\\Console\\Input\\ArgvInput"},{"__class__":"Symfony\\Component\\Console\\Output\\ConsoleOutput"}]},{"file":"/var/www/html/3rdparty/symfony/console/Application.php","line":149,"function":"doRun","class":"Symfony\\Component\\Console\\Application","type":"->","args":[{"__class__":"Symfony\\Component\\Console\\Input\\ArgvInput"},{"__class__":"Symfony\\Component\\Console\\Output\\ConsoleOutput"}]},{"file":"/var/www/html/lib/private/Console/Application.php","line":211,"function":"run","class":"Symfony\\Component\\Console\\Application","type":"->","args":[{"__class__":"Symfony\\Component\\Console\\Input\\ArgvInput"},{"__class__":"Symfony\\Component\\Console\\Output\\ConsoleOutput"}]},{"file":"/var/www/html/console.php","line":99,"function":"run","class":"OC\\Console\\Application","type":"->","args":[]},{"file":"/var/www/html/occ","line":11,"args":["/var/www/html/console.php"],"function":"require_once"}],"File":"/var/www/html/3rdparty/doctrine/dbal/src/Driver/PDO/Exception.php","Line":26,"Previous":{"Exception":"PDOException","Message":"SQLSTATE[23000]: Integrity constraint violation: 1062 Duplicate entry '300' for key 'ldap_user_directory_uuid'","Code":"23000","Trace":[{"file":"/var/www/html/3rdparty/doctrine/dbal/src/Driver/PDO/Connection.php","line":82,"function":"query","class":"PDO","type":"->","args":["CREATE UNIQUE INDEX ldap_user_directory_uuid ON oc_ldap_user_mapping (directory_uuid)"]},{"file":"/var/www/html/3rdparty/doctrine/dbal/src/Connection.php","line":1062,"function":"query","class":"Doctrine\\DBAL\\Driver\\PDO\\Connection","type":"->","args":["CREATE UNIQUE INDEX ldap_user_directory_uuid ON oc_ldap_user_mapping (directory_uuid)"]},{"file":"/var/www/html/lib/private/DB/Connection.php","line":237,"function":"executeQuery","class":"Doctrine\\DBAL\\Connection","type":"->","args":["CREATE UNIQUE INDEX ldap_user_directory_uuid ON oc_ldap_user_mapping (directory_uuid)",[],[],null]},{"file":"/var/www/html/3rdparty/doctrine/dbal/src/Connection.php","line":1809,"function":"executeQuery","class":"OC\\DB\\Connection","type":"->","args":["CREATE UNIQUE INDEX ldap_user_directory_uuid ON oc_ldap_user_mapping (directory_uuid)"]},{"file":"/var/www/html/lib/private/DB/Migrator.php","line":175,"function":"query","class":"Doctrine\\DBAL\\Connection","type":"->","args":["CREATE UNIQUE INDEX ldap_user_directory_uuid ON oc_ldap_user_mapping (directory_uuid)"]},{"file":"/var/www/html/lib/private/DB/Migrator.php","line":76,"function":"applySchema","class":"OC\\DB\\Migrator","type":"->","args":[{"__class__":"Doctrine\\DBAL\\Schema\\Schema"}]},{"file":"/var/www/html/lib/private/DB/Connection.php","line":546,"function":"migrate","class":"OC\\DB\\Migrator","type":"->","args":[{"__class__":"Doctrine\\DBAL\\Schema\\Schema"}]},{"file":"/var/www/html/lib/private/DB/MigrationService.php","line":540,"function":"migrateToSchema","class":"OC\\DB\\Connection","type":"->","args":[{"__class__":"Doctrine\\DBAL\\Schema\\Schema"}]},{"file":"/var/www/html/lib/private/DB/MigrationService.php","line":425,"function":"executeStep","class":"OC\\DB\\MigrationService","type":"->","args":["1130Date20211102154716",false]},{"file":"/var/www/html/lib/private/legacy/OC_App.php","line":1011,"function":"migrate","class":"OC\\DB\\MigrationService","type":"->","args":[]},{"file":"/var/www/html/lib/private/Updater.php","line":353,"function":"updateApp","class":"OC_App","type":"::","args":["*** sensitive parameters replaced ***"]},{"file":"/var/www/html/lib/private/Updater.php","line":265,"function":"doAppUpgrade","class":"OC\\Updater","type":"->","args":[]},{"file":"/var/www/html/lib/private/Updater.php","line":133,"function":"doUpgrade","class":"OC\\Updater","type":"->","args":["23.0.1.2","23.0.0.10"]},{"file":"/var/www/html/core/Command/Upgrade.php","line":235,"function":"upgrade","class":"OC\\Updater","type":"->","args":[]},{"file":"/var/www/html/3rdparty/symfony/console/Command/Command.php","line":255,"function":"execute","class":"OC\\Core\\Command\\Upgrade","type":"->","args":[{"__class__":"Symfony\\Component\\Console\\Input\\ArgvInput"},{"__class__":"Symfony\\Component\\Console\\Output\\ConsoleOutput"}]},{"file":"/var/www/html/3rdparty/symfony/console/Application.php","line":1009,"function":"run","class":"Symfony\\Component\\Console\\Command\\Command","type":"->","args":[{"__class__":"Symfony\\Component\\Console\\Input\\ArgvInput"},{"__class__":"Symfony\\Component\\Console\\Output\\ConsoleOutput"}]},{"file":"/var/www/html/3rdparty/symfony/console/Application.php","line":273,"function":"doRunCommand","class":"Symfony\\Component\\Console\\Application","type":"->","args":[{"__class__":"OC\\Core\\Command\\Upgrade"},{"__class__":"Symfony\\Component\\Console\\Input\\ArgvInput"},{"__class__":"Symfony\\Component\\Console\\Output\\ConsoleOutput"}]},{"file":"/var/www/html/3rdparty/symfony/console/Application.php","line":149,"function":"doRun","class":"Symfony\\Component\\Console\\Application","type":"->","args":[{"__class__":"Symfony\\Component\\Console\\Input\\ArgvInput"},{"__class__":"Symfony\\Component\\Console\\Output\\ConsoleOutput"}]},{"file":"/var/www/html/lib/private/Console/Application.php","line":211,"function":"run","class":"Symfony\\Component\\Console\\Application","type":"->","args":[{"__class__":"Symfony\\Component\\Console\\Input\\ArgvInput"},{"__class__":"Symfony\\Component\\Console\\Output\\ConsoleOutput"}]},{"file":"/var/www/html/console.php","line":99,"function":"run","class":"OC\\Console\\Application","type":"->","args":[]},{"file":"/var/www/html/occ","line":11,"args":["/var/www/html/console.php"],"function":"require_once"}],"File":"/var/www/html/3rdparty/doctrine/dbal/src/Driver/PDO/Connection.php","Line":82}}},"CustomMessage":"Database error when running migration latest for app user_ldap"}}
```

### Additional info

Retrying the upgrade leads to `The table with name 'nextcloud.oc_ldap_group_mapping_backup' already exists`

### PR DESCRIPTION
**Pull request by blizzz**
solves #30936

strategy based on https://github.com/nextcloud/server/issues/30936#issuecomment-1027844910

* [x] invalidate duplicated UUIDs in migration step
* [x] make it possible to re-run migration step for broken setup
* [x] occ command to let Nextcloud try to update UUIDs to current value¹
* [x] add a warning in admin overview if invalidated UUIDs are present
* [x] testing

¹ could be a manually triggerable repair step, but it would run once, be less flexible and this way it MAY be also easier for some LDAP server migration scenarios when an update was needed anyways (but only works of DN stays identical).

### CODE REVIEW

File: apps/user_ldap/lib/Migration/Version1130Date20211102154716.php [hunk 1]
```diff
@@ -52,6 +53,23 @@ public function getName() {
...
+		foreach (['ldap_user_mapping', 'ldap_group_mapping'] as $tableName) {
+			$this->processDuplicateUUIDs($tableName);
+		}
+
+		/** @var ISchemaWrapper $schema */
+		$schema = $schemaClosure();
+		if ($schema->hasTable('ldap_group_mapping_backup')) {
+			// Previous upgrades of a broken release might have left an incomplete
+			// ldap_group_mapping_backup table. No need to recreate, but it
+			// should be empty.
```
Comment (come-nc) [inline apps/user_ldap/lib/Migration/Version1130Date20211102154716.php:66]: My fear here is that depending where the update failed, the backup table may be the only one still having the mapping information. Maybe add an extra check that ldap_group_mapping is existing and not empty (unless backup is empty as well)?

File: apps/user_ldap/lib/Migration/Version1130Date20211102154716.php [hunk 2]
```diff
@@ -172,4 +190,86 @@ protected function getUpdateQuery(string $table): IQueryBuilder {
...
+
+	/**
+	 * @throws Exception
+	 */
+	protected function processDuplicateUUIDs(string $table): void {
+		$uuids = $this->getDuplicatedUuids($table);
+		$idsWithUuidToInvalidate = [];
+		foreach ($uuids as $uuid) {
+			array_push($idsWithUuidToInvalidate, ...$this->getNextcloudIdsByUuid($table, $uuid));
+		}
```
Comment (come-nc) [inline apps/user_ldap/lib/Migration/Version1130Date20211102154716.php:202]: Missing a call to invalidateUuids here I suppose?

File: apps/user_ldap/lib/Migration/Version1130Date20211102154716.php (hunk 1 above)
Comment (blizzz) [reply apps/user_ldap/lib/Migration/Version1130Date20211102154716.php:66]: `ldap_group_mapping` can be legitimately empty when groups are not configured.

Where could this migration step currently fail? 

1. in `changeSchema`, then we would have no backup table at all and therefore this is irrelevant.
2. in `postSchemaChange`, but this affects updating ldap_dn_hashes in the "old" tables only, no deletion or copying.

Therefore, I think we are safe in this context.

Comment (blizzz) [reply apps/user_ldap/lib/Migration/Version1130Date20211102154716.php:66]: I think I did not look into Version1120Date20210917155206 though (discarded if for the lower version number), I shall have a closer look.

Comment (blizzz) [reply apps/user_ldap/lib/Migration/Version1130Date20211102154716.php:66]: Actually truncating is already over-defensive for it should be empty nevertheless. This table is created in this migration step, but not populated yet. The check introduced in `schemaChange` is sufficient, and the truncation block can be dropped entirely.

File: apps/user_ldap/lib/Migration/Version1130Date20211102154716.php (hunk 2 above)
Comment (blizzz) [reply apps/user_ldap/lib/Migration/Version1130Date20211102154716.php:202]: Oh my, I was already a step ahead in my mind.

File: apps/user_ldap/lib/Migration/Version1130Date20211102154716.php (hunk 1 above)
Comment (blizzz) [reply apps/user_ldap/lib/Migration/Version1130Date20211102154716.php]: But… then this migration step should have worked already priorly in #30936 yet, so why would it be executed again? 🤔

**Comment by blizzz** [comment]
Added the checkup warning and did some basic testing with  the occ command, which was working as expected.

Minor annoyance: we have no means to clean up (i.e. remove) old gone groups. That's not a blocker and can be worked around with custom  instructions if necessary.

**Comment by blizzz** [comment]
Rebased and squashes the fixups. My testing was successful so far.

File: apps/user_ldap/lib/Jobs/CleanUp.php [hunk 3]
```diff
@@ -119,38 +119,29 @@ public function setArguments($arguments) {
...
-	public function run($argument) {
+	public function run($argument): void {
 		$this->setArguments($argument);
 
 		if (!$this->isCleanUpAllowed()) {
 			return;
 		}
 		$users = $this->mapping->getList($this->getOffset(), $this->getChunkSize());
-		if (!is_array($users)) {
-			//something wrong? Let's start from the beginning next time and
```
Comment (PVince81) [inline apps/user_ldap/lib/Jobs/CleanUp.php:130]: is this obsolete in general or connected to the context of this PR ?

**Review by PVince81 — COMMENTED** [review]
I flew over the code and didn't find anything notable apart from my minor comments.

I hope others can chime in to review the more complex code more in detail,

File: apps/user_ldap/lib/Command/UpdateUUID.php [hunk 4]
```diff
@@ -0,0 +1,374 @@
...
+use OCA\User_LDAP\User_Proxy;
+use Psr\Log\LoggerInterface;
+use Symfony\Component\Console\Command\Command;
+use Symfony\Component\Console\Helper\ProgressBar;
+use Symfony\Component\Console\Input\InputInterface;
+use Symfony\Component\Console\Input\InputOption;
+use Symfony\Component\Console\Output\OutputInterface;
+use function sprintf;
+
+class UuidUpdateReport {
```
[the remark this answers is not in the document]
Comment (blizzz) [reply apps/user_ldap/lib/Command/UpdateUUID.php:44]: Legal it is. I believe it is also not the sole case in our code bases, but given, I would not know for certain where to look at first. It is an internal helper class that just contains public properties, in order to not use an arbitrary array with with arbitrary keys.

File: apps/user_ldap/lib/Jobs/CleanUp.php (hunk 3 above)
Comment (blizzz) [reply apps/user_ldap/lib/Jobs/CleanUp.php:130]: Obsolete for return type hinting.

**Review by come-nc — APPROVED** [review]
:+1:

**Review by PVince81 — APPROVED** [review]
:+1:

there's a DCO error though

**Comment by PVince81** [comment]
down to 20 because the original fix was put there as well...

**Comment by come-nc** [comment]
I messed up my `git rebase -i` and duplicated two commits from Joas.
Not sure how that happenned but it causes no trouble on master because the commits were already on master.
For instance https://github.com/nextcloud/server/commit/1ab0ed81346f77e4e9faab64f7e02977c2849e1e is a duplicate of https://github.com/nextcloud/server/commit/5f5142ff583dfdc9ae82b01177c839f164bce2dc

I will create the backport by hand for 23 and backport from there.

**Comment by ChristophWurst** [comment]
Command is not documented in admin manual :crying_cat_face:

**Pull request merged**